src/lib/optimizer/strategy/between_composition_rule.cpp

```diff
@@ -0,0 +1,219 @@
+          upper_bound_value_expression->value > boundary.value_expression->value) {
+        upper_bound_value_expression = boundary.value_expression;
+        right_inclusive = true;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::LowerBoundaryInclusive) {
+      if (lower_bound_value_expression == nullptr ||
+          lower_bound_value_expression->value < boundary.value_expression->value) {
+        lower_bound_value_expression = boundary.value_expression;
+        left_inclusive = true;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::UpperBoundaryExclusive) {
+      if (upper_bound_value_expression == nullptr ||
+          upper_bound_value_expression->value <= boundary.value_expression->value) {
+        upper_bound_value_expression = boundary.value_expression;
+        right_inclusive = false;
+        node_scope.push_back(boundary.node);
+      }
+    } else if (boundary.type == ColumnBoundaryType::LowerBoundaryExclusive) {
+      if (lower_bound_value_expression == nullptr ||
+          lower_bound_value_expression->value <= boundary.value_expression->value) {
+        lower_bound_value_expression = boundary.value_expression;
+        left_inclusive = false;
+        node_scope.push_back(boundary.node);
+      }
+    }
+  }
+
+  // Apply boundary check again for the last column
+  if (lower_bound_value_expression != nullptr && upper_bound_value_expression != nullptr) {
+    const auto between_node = PredicateNode::make(
+        std::make_shared<BetweenExpression>(current_column_expression, lower_bound_value_expression,
+                                            upper_bound_value_expression, left_inclusive, right_inclusive));
+    between_nodes.push_back(between_node);
+  } else {
+    predicate_nodes.insert(std::end(predicate_nodes), std::cbegin(node_scope), std::cend(node_scope));
+  }
+
+  predicate_nodes.insert(std::end(predicate_nodes), std::cbegin(between_nodes), std::cend(between_nodes));
+
+  // Insert predicate nodes to LQP
+  // Connect last predicate to chain input
+  predicate_nodes.back()->set_left_input(input);
+
+  // Connect predicates
+  for (size_t predicate_index = 0; predicate_index < predicate_nodes.size() - 1; predicate_index++) {
+    predicate_nodes[predicate_index]->set_left_input(predicate_nodes[predicate_index + 1]);
+  }
+
+  // Connect first predicates to chain output
+  for (size_t output_idx = 0; output_idx < outputs.size(); ++output_idx) {
+    outputs[output_idx]->set_input(input_sides[output_idx], predicate_nodes.front());
+  }
+}
+
+void BetweenCompositionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  if (node->type == LQPNodeType::Predicate) {
+    std::vector<std::shared_ptr<AbstractLQPNode>> predicate_nodes;
+
+    // Gather adjacent PredicateNodes
+    auto current_node = node;
+    while (current_node->type == LQPNodeType::Predicate) {
```

Comment:
Probably out of scope, just wondering if the following scenario can happen:
Consider the following sequence of LQP nodes (Predicate `A > 10`), (Projection `A`),(Predicate `A < 20`).
If we get it right, such a case would not be optimized by your rule, allthough its possible in this specific case. Do you know by chance if the given sequence is actually valid, i.e. can it appear in that order or are there some optimizer/sql-parser-intern assumptions that ensure predicate chains are as long as possible? (Not saying you have to handle this)

Reply:
In that case the given scenario will (probably) not occur. I think we can mark this as resolved then?